File: archive/cdk-py/Pr6437/lib/lambda_stack.py
from aws_cdk import (
    Stack,
    aws_lambda as _lambda,
    aws_ec2 as ec2,
    aws_iam as iam,
    Duration,
    Tags,
    CfnOutput
)
from constructs import Construct
import os

class LambdaStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, vpc: ec2.Vpc,
                 environment_suffix: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Lambda execution role
        lambda_role = iam.Role(
            self, f"LambdaExecutionRole-{environment_suffix}",
            assumed_by=iam.ServicePrincipal("lambda.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name("service-role/AWSLambdaVPCAccessExecutionRole"),
                iam.ManagedPolicy.from_aws_managed_policy_name("service-role/AWSLambdaBasicExecutionRole")
            ]
        )

        # Grant DynamoDB permissions
        lambda_role.add_to_policy(
            iam.PolicyStatement(
                actions=[
                    "dynamodb:PutItem",
                    "dynamodb:GetItem",
                    "dynamodb:UpdateItem",
                    "dynamodb:Query",
                    "dynamodb:Scan"
                ],
                resources=[f"arn:aws:dynamodb:{self.region}:{self.account}:table/SessionTable-{environment_suffix}"]
            )
        )

        # Grant SNS permissions
        lambda_role.add_to_policy(
            iam.PolicyStatement(
                actions=["sns:Publish"],
                resources=["*"]
            )
        )

        # Payment Validation Lambda
        self.payment_validation_fn = _lambda.Function(
            self, f"PaymentValidation-{environment_suffix}",
            function_name=f"payment-validation-{environment_suffix}",
            runtime=_lambda.Runtime.PYTHON_3_11,
            handler="index.handler",
            code=_lambda.Code.from_asset("lib/lambda/payment_validation"),
            role=lambda_role,
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS),
            timeout=Duration.seconds(30),
            memory_size=256,
            environment={
                "ENVIRONMENT_SUFFIX": environment_suffix
            }
        )

        # Transaction Processing Lambda
        self.transaction_processing_fn = _lambda.Function(
            self, f"TransactionProcessing-{environment_suffix}",
            function_name=f"transaction-processing-{environment_suffix}",
            runtime=_lambda.Runtime.PYTHON_3_11,
            handler="index.handler",
            code=_lambda.Code.from_asset("lib/lambda/transaction_processing"),
            role=lambda_role,
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS),
            timeout=Duration.seconds(30),
            memory_size=256,
            environment={
                "ENVIRONMENT_SUFFIX": environment_suffix
            }
        )

        # Notification Lambda
        self.notification_fn = _lambda.Function(
            self, f"Notification-{environment_suffix}",
            function_name=f"notification-{environment_suffix}",
            runtime=_lambda.Runtime.PYTHON_3_11,
            handler="index.handler",
            code=_lambda.Code.from_asset("lib/lambda/notification"),
            role=lambda_role,
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS),
            timeout=Duration.seconds(30),
            memory_size=256,
            environment={
                "ENVIRONMENT_SUFFIX": environment_suffix
            }
        )

        # Outputs
        CfnOutput(
            self, "PaymentValidationFnArn",
            value=self.payment_validation_fn.function_arn,
            export_name=f"payment-validation-fn-{environment_suffix}"
        )

        CfnOutput(
            self, "TransactionProcessingFnArn",
            value=self.transaction_processing_fn.function_arn,
            export_name=f"transaction-processing-fn-{environment_suffix}"
        )

        CfnOutput(
            self, "NotificationFnArn",
            value=self.notification_fn.function_arn,
            export_name=f"notification-fn-{environment_suffix}"
        )

File: lib/stacks/lambda-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';
import { BaseStack, BaseStackProps } from './base-stack';
import * as ec2 from 'aws-cdk-lib/aws-ec2';

export interface LambdaStackProps extends BaseStackProps {
  vpc: ec2.IVpc;
}

export class LambdaStack extends BaseStack {
  public readonly orderProcessingFunction: lambda.Function;

  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id, props);

    // Create execution role with environment-specific permissions boundary
    const executionRole = new iam.Role(this, 'OrderProcessingRole', {
      roleName: this.getResourceName('order-processing-role'),
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AWSLambdaVPCAccessExecutionRole'
        ),
      ],
    });

    // Add least-privilege permissions
    executionRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: [
          'dynamodb:PutItem',
          'dynamodb:GetItem',
          'dynamodb:UpdateItem',
          'dynamodb:Query',
        ],
        resources: [
          `arn:aws:dynamodb:${this.region}:${this.account}:table/orders-${this.environmentSuffix}`,
        ],
      })
    );

    executionRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['sqs:SendMessage', 'sqs:GetQueueUrl'],
        resources: [
          `arn:aws:sqs:${this.region}:${this.account}:order-processing-${this.environmentSuffix}`,
        ],
      })
    );

    executionRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['s3:PutObject', 's3:GetObject'],
        resources: [`arn:aws:s3:::trade-data-${this.environmentSuffix}/*`],
      })
    );

    // Security group for Lambda
    const lambdaSecurityGroup = new ec2.SecurityGroup(
      this,
      'LambdaSecurityGroup',
      {
        vpc: props.vpc,
        securityGroupName: this.getResourceName('lambda-sg'),
        description: 'Security group for order processing Lambda',
        allowAllOutbound: true,
      }
    );

    // Create Lambda function with environment-specific configuration
    this.orderProcessingFunction = new lambda.Function(
      this,
      'OrderProcessingFunction',
      {
        functionName: this.getResourceName('order-processing'),
        runtime: lambda.Runtime.NODEJS_18_X,
        handler: 'index.handler',
        code: lambda.Code.fromAsset('lib/lambda/order-processing'),
        memorySize: this.environmentConfig.lambdaConfig.memorySize,
        timeout: cdk.Duration.seconds(
          this.environmentConfig.lambdaConfig.timeout
        ),
        reservedConcurrentExecutions:
          this.environmentConfig.lambdaConfig.reservedConcurrentExecutions,
        role: executionRole,
        vpc: props.vpc,
        vpcSubnets: {
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        },
        securityGroups: [lambdaSecurityGroup],
        environment: {
          ENVIRONMENT: this.environmentSuffix,
          DYNAMODB_TABLE: `orders-${this.environmentSuffix}`,
          SQS_QUEUE: `order-processing-${this.environmentSuffix}`,
          S3_BUCKET: `trade-data-${this.environmentSuffix}`,
        },
        logRetention: logs.RetentionDays.ONE_MONTH,
        tracing: lambda.Tracing.ACTIVE,
      }
    );

    // Export Lambda function ARN
    this.exportToParameterStore(
      'order-processing-function-arn',
      this.orderProcessingFunction.functionArn
    );
  }
}
